probe-rs/src/flashing/progress.rs

```diff
@@ -0,0 +1,128 @@
+/// A structure to manage the flashing procedure progress reporting.
+///
+/// This struct stores a handler closure which will be called everytime an event happens during the flashing process.
+/// Such an event can be start or finish of the flashing procedure or a progress report, as well as some more events.
+pub struct FlashProgress {
+    handler: Box<dyn Fn(ProgressEvent)>,
+}
+
+impl FlashProgress {
+    /// Create a new `FlashProgress` structure with a given `handler` to be called on events.
+    pub fn new(handler: impl Fn(ProgressEvent) + 'static) -> Self {
+        Self {
+            handler: Box::new(handler),
+        }
+    }
+
+    /// Emit a flashing progress event.
+    fn emit(&self, event: ProgressEvent) {
+        (self.handler)(event);
+    }
+
+    /// Signalize that the flashing algorithm was set up and is initialized.
+    pub(super) fn initialized(
+        &self,
+        total_pages: usize,
+        total_sector_size: usize,
+        page_size: u32,
+        fill_size: u32,
+    ) {
+        self.emit(ProgressEvent::Initialized {
+            total_pages,
+            total_sector_size,
+            page_size,
+            fill_size,
+        });
+    }
+
+    /// Signalize that the entire flashing procedure started.
+    pub(super) fn started_flashing(&self) {
+        self.emit(ProgressEvent::StartedFlashing);
+    }
+
+    /// Signalize that the erasing procedure started.
+    pub(super) fn started_erasing(&self) {
+        self.emit(ProgressEvent::StartedErasing);
+    }
+
+    /// Signalize that the filling procedure started.
+    pub(super) fn started_filling(&self) {
+        self.emit(ProgressEvent::StartedFilling);
+    }
+
+    /// Signalize that the page programming procedure has made progress.
+    pub(super) fn page_programmed(&self, size: u32, time: u128) {
+        self.emit(ProgressEvent::PageFlashed { size, time });
+    }
+
+    /// Signalize that the sector erasing procedure has made progress.
+    pub(super) fn sector_erased(&self, size: u32, time: u128) {
+        self.emit(ProgressEvent::SectorErased { size, time });
+    }
+
+    /// Signalize that the page filling procedure has made progress.
+    pub(super) fn page_filled(&self, size: u32, time: u128) {
```

Comment:
How long does it take to fill a page? I'm wondering if this isn't too short to be noticeable.

Reply:
It needs to read a max of page_size bytes from flash. I think this is a noticeable timespan. and even if it is not for a single piece it is for multiple of those and it's quite common that more than a full page has to be filled.